---
title: "Analysis of First LMMP scRNAseq on 6 month old male murine ileum"
author: "Sushma Nagaraj"
date: "6/18/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
```

```{r init}
library(monocle3)
library(here)
library(scater)
library(scran)
library(dplyr)
#library(SingleCellExperiment)
library(AnnotationHub)
library(DropletUtils)
library(pheatmap)
```

# Create SCE object
```{r sce_object}
cds.macrophage <- readRDS(here("data", "Kulkarni_cds.rds")) #Monocle CellDataSet object 
genes <- fData(cds.macrophage)[,c("id","symbol")]
mycolumns <- c("barcode","treatment","total_UMI", "num_genes_detected", "cell_type")
pheno <- pData(cds.macrophage)[,mycolumns]
counts <- exprs(cds.macrophage)
sce.macrophage <- SingleCellExperiment(assays = list(counts = counts), colData = pheno, rowData = genes)

ah <- AnnotationHub()
ens.mm.v97 <- ah[['AH73905']]
location <- mapIds(ens.mm.v97, keys = rownames(sce.macrophage), keytype = "GENEID", column = "SEQNAME")

dups <- duplicated(rownames(sce.macrophage)) #No duplicate entries found..
dups <- duplicated(rowData(sce.macrophage)$symbol)
rownames(sce.macrophage) <- uniquifyFeatureNames(rowData(sce.macrophage)$id, rowData(sce.macrophage)$symbol)

rowData(sce.macrophage)$chr <- location 
```

# QC
```{r QC}
bcrank <- barcodeRanks(counts(sce.macrophage))
uniq <- !duplicated(bcrank$rank)
pdf(file = here("plots", "Knee_plot.pdf"))
plot(bcrank$rank[uniq], bcrank$total[uniq], log="xy", xlab = "Rank", ylab = "Total UMI count", cex.lab = 1.2)
abline(h=metadata(bcrank)$inflection, col = "darkgreen", lty = 2)
abline(h=metadata(bcrank)$knee, col = "dodgerblue", lty = 2)
legend("bottomleft", legend = c("Inflection", "Knee"), col = c("darkgreen", "dodgerblue"), lty = 2, cex = 1.2)
dev.off()

mito <- which(rowData(sce.macrophage)$chr == "MT")
stats <- perCellQCMetrics(sce.macrophage, subsets = list(Mt=mito))
qc <- quickPerCellQC(stats, percent_subsets = "subsets_Mt_percent", batch = sce.macrophage$treatment)
colSums(as.matrix(qc))
#low_lib_size          low_n_features high_subsets_Mt_percent 
#0                     403                     274 
#discard 
#643 

colData(sce.macrophage) <- cbind(colData(sce.macrophage), stats)
sce.macrophage$treatment <- factor(sce.macrophage$treatment)
sce.macrophage$discard <- qc$discard
pdf(here("plots", "QC_violin_plots.pdf"))
gridExtra::grid.arrange(
  plotColData(sce.macrophage, x="treatment", y="sum", colour_by = "discard") + scale_y_log10() + ggtitle("Total count"),
  plotColData(sce.macrophage, x="treatment", y="detected", colour_by = "discard") + scale_y_log10() + ggtitle("Detected features"),
  plotColData(sce.macrophage, x="treatment", y="subsets_Mt_percent", colour_by = "discard") + ggtitle("Mito percent"),
  nrow = 2, ncol = 2
)
dev.off()
discard.mito <- isOutlier(sce.macrophage$subsets_Mt_percent, type = "higher", batch = sce.macrophage$treatment)
attr(discard.mito, "thresholds")
#TH       TL
#lower      -Inf     -Inf
#higher 8.876797 12.43833
colData(sce.macrophage)$discard.mito <- discard.mito

discard.features <- isOutlier(sce.macrophage$detected, type = "lower", log = TRUE, batch = sce.macrophage$treatment)
attr(discard.features, "thresholds")
#TH       TL
#lower  405.6947 325.1462
#higher      Inf      Inf
colData(sce.macrophage)$discard.features <- discard.features

discard.sum <- isOutlier(sce.macrophage$sum, type = "lower", log = TRUE, batch = sce.macrophage$treatment)
attr(discard.sum, "thresholds")
#TH       TL
#lower  585.8507 653.1038
#higher      Inf      Inf

pdf(here("plots", "Counts-vs-MitoPercent.pdf"))
plotColData(sce.macrophage, x = "sum", y = "subsets_Mt_percent", colour_by = "discard") + scale_x_log10()
dev.off()

pdf(here("plots", "QC_scatterplots.2_populations.pdf"))
gridExtra::grid.arrange(                       
  plotColData(sce.macrophage, x = "sum", y = "detected", colour_by = "discard.features") + scale_x_log10() + scale_y_log10(),
  plotColData(sce.macrophage, x = "sum", y = "detected", colour_by = "discard.mito") + scale_x_log10() + scale_y_log10(),
  plotColData(sce.macrophage, x = "detected", y = "subsets_Mt_percent", colour_by = "discard.features") + scale_x_log10(),
  plotColData(sce.macrophage, x = "detected", y = "subsets_Mt_percent", colour_by = "discard.mito") + scale_x_log10(),
  nrow = 2,
  ncol = 2
)
dev.off()

pdf(here("plots", "QC_Sum-vs-Detected.smoothscatter.pdf"))
smoothScatter(log10(sce.macrophage$sum), log10(sce.macrophage$detected), nbin = 250)
dev.off()

pdf(here("plots", "QC_Detected-vs-MitoPercent.pdf"))
plotColData(sce.macrophage, x = "detected", y = "subsets_Mt_percent", colour_by = "treatment", shape_by = "discard.mito") + scale_x_log10()
dev.off()

unfiltered <- sce.macrophage
sce.macrophage <- sce.macrophage[,!sce.macrophage$discard]
```

# Normalization
```{r normalization}
set.seed(1000)
clusters <- quickCluster(sce.macrophage)
table(clusters)
sce.macrophage <- computeSumFactors(sce.macrophage, cluster = clusters)
sce.macrophage <- logNormCounts(sce.macrophage)
summary(sizeFactors(sce.macrophage))

pdf(here("plots", "Normalization.Library_size_factors-vs-Deconvolution_factors.pdf"))
plot(librarySizeFactors(sce.macrophage), sizeFactors(sce.macrophage), pch = 16, xlab = "Library size factors", ylab = "Deconvolution factors", log = "xy")
dev.off()
```
 
#Variance modelling (dec = decompose)
```{r model_var}
dec.macrophage <- modelGeneVar(sce.macrophage, block = sce.macrophage$treatment)
chosen.hvgs <- getTopHVGs(dec.macrophage, prop = 0.1)
par(mfrow = c(1,2))
blocked.stats <- dec.macrophage$per.block
pdf(here("plots", "Variance_modelling.pdf"))
for (i in colnames(blocked.stats)) {
  current <- blocked.stats[[i]]
  plot(current$mean, current$total, main = i, pch = 16, cex = 0.5, xlab = "Mean of log-expression", ylab = "Variance of log-expression")
  curfit <- metadata(current)
  curve(curfit$trend(x), col = "dodgerblue", add = TRUE, lwd = 2)
}
dev.off()
```

#Batch effect detection
```{r batch_effects}
hvgs <- dec.macrophage$bio > 0

set.seed(0010101010)
uncorrected <- runPCA(sce.macrophage, subset_row = hvgs, BSPARAM = BiocSingular::RandomParam())
snn.gr.uncorr <- buildSNNGraph(uncorrected, use.dimred = "PCA")
clusters.uncorr <- igraph::cluster_walktrap(snn.gr.uncorr)$membership
tab.uncorr <- table(Cluster = clusters.uncorr, Batch = uncorrected$treatment)
colData(uncorrected)$clusters <- clusters.uncorr
tab

set.seed(1111001)
uncorrected <- runTSNE(uncorrected, dimred = "PCA")
pdf(here("plots", "Batch_correction.uncorrected_Batches.pdf"))
plotTSNE(uncorrected, colour_by = "treatment")
dev.off()

#Batch Correction using rescaleBatches
uncorrected <- sce.macrophage
library(batchelor)
corrected <- rescaleBatches(sce.macrophage, assay.type = "logcounts", batch = sce.macrophage$treatment)
#assay(sce.macrophage, "corrected") <- assay(corrected, "corrected")

set.seed(0010101010)
corrected <- runPCA(corrected, subset_row = hvgs, exprs_values = "corrected", BSPARAM = BiocSingular::RandomParam())
snn.gr <- buildSNNGraph(corrected, use.dimred = "PCA")
clusters <- igraph::cluster_walktrap(snn.gr)$membership
tab <- table(Cluster = clusters, Batch = uncorrected$treatment)
tab

set.seed(1111001)
corrected <- runTSNE(corrected, dimred = "PCA")
pdf(here("plots", "Batch_correction.rescaleBatches.pdf"))
plotTSNE(corrected, colour_by = "treatment")
dev.off()

#Batch correction using fastMNN and diagnostics.
set.seed(1000101001)
mnn.out <- fastMNN(sce.macrophage, subset.row = hvgs, batch = sce.macrophage$treatment, BSPARAM = BiocSingular::RandomParam(deferred = TRUE))
#mnn.out <- fastMNN(sce.macrophage, subset.row = chosen.hvgs, batch = sce.macrophage$treatment, BSPARAM = BiocSingular::RandomParam(deferred = TRUE))
dim(reducedDim(mnn.out, "corrected"))
snn.gr <- buildSNNGraph(mnn.out, use.dimred = "corrected")
clusters.mnn <- igraph::cluster_walktrap(snn.gr)$membership
tab.mnn <- table(Cluster = clusters.mnn, Batch = mnn.out$batch)
tab.mnn

set.seed(0010101010)
mnn.out <- runTSNE(mnn.out, dimred = "corrected")
mnn.out$batch <- factor(mnn.out$batch)
pdf(here("plots", "Batch_correction.fastMNN.pdf"))
plotTSNE(mnn.out, colour_by = "batch")
dev.off()

metadata(mnn.out)$merge.info$lost.var
#             TH         TL
#[1,] 0.02763112 0.04459383

library(pheatmap)
tab_before_after <- table(paste0("after_", clusters.mnn[sce.macrophage$treatment == "TH"]), paste0("before_", clusters.uncorr[sce.macrophage$treatment == "TH"]))
heatTH <- pheatmap(log10(tab_before_after+10), cluster_rows = FALSE, cluster_cols = FALSE, main = "TH comparison", silent = TRUE)

tab_before_after_2 <- table(paste0("after_", clusters.mnn[sce.macrophage$treatment == "TL"]), paste0("before_", clusters.uncorr[sce.macrophage$treatment == "TL"]))
heatTL <- pheatmap(log10(tab_before_after_2+10), cluster_rows = FALSE, cluster_cols = FALSE, main = "TL comparison", silent = TRUE)

pdf(here("plots", "Comparison_of_clusters_before_and_after_MNNcorrection.pdf"))
gridExtra::grid.arrange(heatTH[[4]], heatTL[[4]])
dev.off()
```

# Batch correction using fastMNN and dimensionality reduction
```{r reduce_dims}
set.seed(1000101001)
mnn.out <- fastMNN(sce.macrophage, subset.row = chosen.hvgs, batch = sce.macrophage$treatment, BSPARAM = BiocSingular::RandomParam(deferred = TRUE))
dim(reducedDim(mnn.out, "corrected"))
altExp(sce.macrophage, "mnn.out") <- mnn.out
reducedDim(sce.macrophage, "corrected") <- reducedDim(mnn.out, "corrected")
set.seed(1010)
sce.macrophage <- runTSNE(sce.macrophage, dimred = "corrected")
tSNE_plot <- plotTSNE(sce.macrophage, colour_by = "treatment")

set.seed(1100101001)
sce.macrophage <- runUMAP(sce.macrophage, dimred = "corrected")
umap_plot <- plotReducedDim(sce.macrophage, dimred = "UMAP", colour_by = "treatment")

pdf(here("plots", "TSNE_UMAP.bybatch.pdf"))
gridExtra::grid.arrange(tSNE_plot, umap_plot, ncol = 2)
dev.off()
```

# Clustering
```{r clustering}
g.10 <- buildSNNGraph(sce.macrophage, k=10, use.dimred = "corrected")
clust.10 <- igraph::cluster_louvain(g.10)$membership
table(clust.10)
sce.macrophage$clust.10 <- factor(clust.10)
umap.2 <- plotUMAP(sce.macrophage, colour_by = "clust.10")
tsne.2 <- plotTSNE(sce.macrophage, colour_by = "clust.10")

pdf(here("plots", "TSNE_UMAP.bycluster.pdf"))
gridExtra::grid.arrange(tsne.2, umap.2, ncol = 2)
dev.off()

ratio <- clusterModularity(g.10, clust.10, as.ratio = TRUE)
dim(ratio)

pdf(here("plots", "cluster_separation.pdf"))
pheatmap(log2(ratio+1), cluster_rows = FALSE, cluster_cols = FALSE, color = colorRampPalette(c("white", "blue"))(100))
dev.off()
```

# Cell type annotation
```{r annotation}
library(SingleR)
ref <- MouseRNAseqData()
pred <- SingleR(test = sce.macrophage, ref = ref, labels = ref$label.main)
table(pred$labels)

pdf(here("plots","Cell_type_annotation_score.heatmap.pdf"))
plotScoreHeatmap(pred)
dev.off()

pdf(here("plots", "Comparison_of_annotation_to_clusters.heatmap.pdf"))
tab <- table(Assigned = pred$pruned.labels, Cluster = sce.macrophage$clust.10)
pheatmap(log2(tab+10), color = colorRampPalette(c("white", "blue"))(101))
dev.off()

colData(sce.macrophage)$cell_type_ann <- pred$pruned.labels

pdf(here("plots", "tSNE.cell_type_annotation.pdf"))
plotTSNE(sce.macrophage, colour_by = "cell_type_ann")
dev.off()

table(sce.macrophage$clust.10, sce.macrophage$cell_type_ann)
```

# Marker genes
```{r marker_genes}
markers <- findMarkers(sce.macrophage, sce.macrophage$clust.10, block = sce.macrophage$treatment, BPPARAM = BiocParallel::MulticoreParam(1))
markers[["6"]][1:5]

genes <- lapply(markers, function(x) {rownames(x)[x$Top <= 5]})
genes <- unique(unlist(genes))

pdf(here("plots", "markers_heatmap.pdf"))
plotHeatmap(sce.macrophage, genes,
            colour_columns_by = c("clust.10", "treatment"),
            show_colnames = FALSE,
            clustering_method = 'ward.D2',
            fontsize_row = 4)
dev.off()

markers <- findMarkers(sce.macrophage, sce.macrophage$clust.10, block = sce.macrophage$treatment, direction = "up", BPPARAM = BiocParallel::MulticoreParam(1))
genes <- lapply(markers, function(x) {rownames(x)[x$Top <= 20]})
sink("genes_per_cluster.txt")
unlist(lapply(genes, paste, collapse = " "))
sink()

plotExpression(sce.macrophage, features=c("Cd14", "Cd68"), x=I(factor(sce.macrophage$clust.10)), colour_by=I(factor(sce.macrophage$clust.10)))
plotTSNE(sce.macrophage, colour_by = "Cd74") #Macrophage
plotTSNE(sce.macrophage, colour_by = "Cd63")
plotTSNE(sce.macrophage, colour_by = "Fcgr3")
```

## Manual annotation and plot.
#"MENs", "NENs_&_GLIA", "VSMC_&_Endothelium", "T_&_NK_Lymphocytes", "Pericytes", "Macrophage_1", "B_Lymphocytes", "Mesothelium", "Macrophage_2", "Macrophage_3", "Macrophage_4", "Adipocytes"
```{r manual_ann}
library(dplyr)
mydata <- data.frame(clust.10 = sce.macrophage$clust.10)
mydata <- mydata %>% mutate(manual_ann = case_when(clust.10 == '1' ~ "MENs_1", 
                                                       clust.10 == '2' ~ "NENs_&_GLIA", 
                                                       clust.10 == '3' ~ "VSMC_&_Endothelium", 
                                                       clust.10 == '4' ~ "T_&_NK_Lymphocytes", 
                                                       clust.10 == '5' ~ "Pericytes", 
                                                       clust.10 == '6' ~ "Macrophage_1", 
                                                       clust.10 == '7' ~ "B_Lymphocytes", 
                                                       clust.10 == '8' ~ "MENs_2", 
                                                       clust.10 == '9' ~ "Macrophage_2", 
                                                       clust.10 == '10' ~ "Macrophage_3", 
                                                       clust.10 == '11' ~ "Macrophage_4", 
                                                       clust.10 == '12' ~ "Adipocytes"))

colData(sce.macrophage)$manual_ann <- mydata$manual_ann

annot <- character()
for (i in unique(sce.macrophage$clust.10)){
  annot <- c(annot, colData(sce.macrophage)[sce.macrophage$clust.10 == i, "barcode"] %>% head(1) %>% as.character())
}

mydat <- as.data.frame(reducedDim(sce.macrophage, "TSNE"))
mydat$label <- paste0(sce.macrophage$clust.10, ".", sce.macrophage$manual_ann)
mydat$cell_type <- sce.macrophage$manual_ann

library(ggrepel)
library(RColorBrewer)

pal <- brewer.pal(12, "Paired")
pdf(here("plots", "TSNE.Manual_annotation.pdf"))
ggplot(data = mydat, aes(x = V1, y = V2, color = cell_type)) + geom_point(size = 1.5, alpha = .4, shape = 21, color = "darkgrey", aes(fill = cell_type)) + 
  scale_color_manual(values = pal, name = "Manual Annotation") + scale_fill_manual(values = pal, name = "Manual Annotation") +
  theme_classic() +
  geom_text_repel(
    data = mydat[annot, ],
    aes(label = label),
    size = 3,
    color = "black"
  )
dev.off()
```

## Doublet detection
```{r doublets}
dbl.out <- doubletCluster(sce.macrophage, sce.macrophage$clust.10)
dbl.out
write.table(dbl.out, file = "doubletCluster.out.txt")
library(BiocSingular)
set.seed(1000101001)
dbl.dens <- doubletCells(sce.macrophage, subset.row = chosen.hvgs, d = ncol(colData(sce.macrophage, "correced")), BPPARAM = BiocParallel::MulticoreParam(1))
summary(dbl.dens)
sce.macrophage$DoubletScore <- log10(dbl.dens+1)
plotTSNE(sce.macrophage, colour_by = "DoubletScore")
pdf(here("plots", "Doublet_scores.pdf"))
plotColData(sce.macrophage, x = "clust.10", y = "DoubletScore", colour_by = "clust.10")
dev.off()
```

#Subset Macrophages and find differentially expressed genes.
```{r}
sce.subset <- sce.macrophage[,sce.macrophage$clust.10 %in% c(6, 9, 10, 11)]
markers <- findMarkers(sce.subset, sce.subset$clust.10, block = sce.subset$treatment, BPPARAM = BiocParallel::MulticoreParam(1))
length(markers)

genes <- lapply(markers, function(x) {rownames(x)[x$Top <= 20]})
genes <- unique(unlist(genes))
genes <- genes[!is.na(genes)]

pdf(here("plots", "macrophage_clusters_markers_heatmap.pdf"))
plotHeatmap(sce.subset, genes,
            colour_columns_by = c("clust.10", "treatment"),
            show_colnames = FALSE,
            clustering_method = 'ward.D2',
            fontsize_row = 4)
dev.off()
```

#Subset Clusters 1 and 8 and find differentially expressed genes.
```{r}
sce.subset <- sce.macrophage[,sce.macrophage$clust.10 %in% c(1, 8)]
markers <- findMarkers(sce.subset, sce.subset$clust.10, block = sce.subset$treatment, BPPARAM = BiocParallel::MulticoreParam(1))

genes <- lapply(markers, function(x) {rownames(x)[x$Top <= 20]})
genes <- unique(unlist(genes))
genes <- genes[!is.na(genes)]
pdf("clusters_1_and_8_markers_heatmap.pdf")
plotHeatmap(sce.subset, genes,
            colour_columns_by = c("clust.10", "treatment"),
            show_colnames = FALSE,
            clustering_method = 'ward.D2',
            fontsize_row = 6)
dev.off()
```

# Cluster 11
## These genes are not exclusively expressed by cluster 11. 
```{r}
plotExpression(sce.subset, features = c("Sparc", "Fxyd1", "Prnp", "Gpm6b"), x = I(as.factor(sce.subset$clust.10)), colour_by = I(as.factor(sce.subset$clust.10)))
plotTSNE(sce.macrophage, colour_by = "Sparc") + scale_fill_gradient(low = "#0072B2", high = "red", name = "Sparc")
plotTSNE(sce.macrophage, colour_by = "Fxyd1") + scale_fill_gradient(low = "#0072B2", high = "red", name = "Fxyd1")
plotTSNE(sce.macrophage, colour_by = "Prnp") + scale_fill_gradient(low = "#0072B2", high = "red", name = "Prnp")
plotTSNE(sce.macrophage, colour_by = "Gpm6b") + scale_fill_gradient(low = "#0072B2", high = "red", name = "Gpm6b")
plotTSNE(sce.macrophage, colour_by = "Csrp1") + scale_fill_gradient(low = "#0072B2", high = "red", name = "Csrp1")
plotTSNE(sce.macrophage, colour_by = "Lsp1") + scale_fill_gradient(low = "#0072B2", high = "red", name = "Lsp1")
plotTSNE(sce.macrophage, colour_by = "F3") + scale_fill_gradient(low = "#0072B2", high = "red", name = "F3")

plotTSNE(sce.subset, colour_by = "L1cam") + scale_fill_gradient(low = "#0072B2", high = "red", name = "L1cam")
plotTSNE(sce.subset, colour_by = "Cdc5l") + scale_fill_gradient2(low = "#0072B2", high = "red", name = "Cdc5l")
plotTSNE(sce.subset, colour_by = "Cd163") + scale_fill_gradient2(low = "#0072B2", high = "red", name = "Cd163")
plotTSNE(sce.subset, colour_by = "Ccr2") + scale_fill_gradient2(low = "#0072B2", high = "red", name = "Ccr2")
plotTSNE(sce.subset, colour_by = "clust.10")

plotExpression(sce.subset, features = c("Lsp1", "Tspan32", "1700017B05Rik", "Irak2", "Ccr2", "Cd163"), x = I(as.factor(sce.subset$clust.10)), colour_by = I(as.factor(sce.subset$clust.10)))
pdf("Cluster11/Expression_Ccr2_Cd163_Lsp1.pdf")
plotExpression(sce.subset, features = c("Lsp1", "Tspan32", "1700017B05Rik", "Irak2", "Ccr2", "Cd163"), x = I(as.factor(sce.subset$clust.10)), colour_by = I(as.factor(sce.subset$clust.10)))
dev.off()


c11 <- markers[[11]]
c11 <- c11[,-c(4,5,6,7,8,10,11,14)]
top <- c11[c11$Top <= 100, ]
write.table(top, file = "Cluster11/focused_DE.top100_genes.txt")

markers.wilcox <- findMarkers(sce.subset, sce.subset$clust.10, block = sce.subset$treatment, test = "wilcox", BPPARAM = BiocParallel::MulticoreParam(1))
c11.w <- markers.wilcox[[11]]
c11.w <- c11.w[,-c(4,5,6,7,8,10,11,14)]
top.w <- c11.w[c11.w$Top <= 100, ]
write.table(top.w, file = "Cluster11/focused_DE.wilcox.top100_genes.txt")

markers.bin <- findMarkers(sce.subset, sce.subset$clust.10, block = sce.subset$treatment, test = "binom", BPPARAM = BiocParallel::MulticoreParam(1))
c11.b <- markers.bin[[11]]
c11.b <- c11.b[,-c(4,5,6,7,8,10,11,14)]
top.b <- c11.b[c11.b$Top <= 100, ]
write.table(top.b, file = "Cluster11/focused_DE.binomial.top100_genes.txt")
```

#Save SCE
```{r save}
saveRDS(sce.macrophage, file = here("data", "sce_Macrophage.rds"))
```

# Cellphonedb
```{r cellphonedb}
library(biomaRt)
human <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", "dataset" = "hsapiens_gene_ensembl", host = "uswest.ensembl.org")
mouse <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", "dataset" = "mmusculus_gene_ensembl", host = "uswest.ensembl.org")
mouseIDs <- rowData(sce.macrophage)[,"id"]
human_orthologs <- getLDS(attributes = c("ensembl_gene_id"), filters = "ensembl_gene_id", values = mouseIDs, mart = mouse, attributesL = c("ensembl_gene_id"), martL = human, uniqueRows = TRUE)
colnames(human_orthologs) <- c("mouseIDs", "humanIDs")
temp <- duplicated(human_orthologs$mouseIDs)
human_orthologs1 <- human_orthologs[!temp,]
mouseIDs <- as.data.frame(mouseIDs, stringsAsFactors = FALSE)
mapping <- left_join(mouseIDs, human_orthologs1)

mycounts <- logcounts(sce.macrophage)
#genes <- data.frame("Gene" = row.names(mycounts))
genes <- data.frame(Gene = mapping$humanIDs)
mycounts <- cbind(genes, mycounts)
na_rows <- is.na(mycounts$Gene)
mycounts <- mycounts[!na_rows, ]
write.table(mycounts, file = here("Macrophage", "cellphonedb", "count.txt"), row.names = FALSE)

myannotation <- data.frame(Cell = row.names(colData(sce.macrophage)), cell_type = colData(sce.macrophage)[,"manual_ann"])
write.table(myannotation, file = here("Macrophage", "cellphonedb", "meta.txt"), row.names = FALSE)
```

# Functions
```{r functions}
getMarkerEffects <- function(x, prefix="logFC", strip=TRUE, remove.na.col=FALSE) {
  regex <- paste0("^", prefix, "\\.")
  i <- grep(regex, colnames(x))
  out <- as.matrix(x[,i])
  
  if (strip) {
    colnames(out) <- sub(regex, "", colnames(out))
  }
  if (remove.na.col) {
    out <- out[,!colAnyNAs(out),drop=FALSE]
  }
  
  out
}
```
